From e0ca9590c7c0be2c80c69a9711b920400c0d89a6 Mon Sep 17 00:00:00 2001
From: Sebastien MICHEL <sebastien.michel@intel.com>
Date: Fri, 7 Sep 2012 18:22:08 +0200
Subject: Fix Biggest Memory leaks on Browser application

BZ: 56189

Working on Issue 31125 (http://code.google.com/p/android/issues
/detail?id=31125):
* Enhance bitmap management (add recycle() functions when necessary)
* Enhance mTabViews management on NavScreen class (add remove()
function when necessary).

Category: AOSP Improvement
Domain: AOSP-Browser
Origin: Internal
Upstream-Candidate: no, key fix

Change-Id: Ic040fec6ebeb0193c97aee3606d629f2c75339c1
Orig-MCG-Change-Id: I04c1802dffe72b9343badb1e7cdee9e1e96f6438
Signed-off-by: Sebastien MICHEL <sebastien.michel@intel.com>
---
 src/com/android/browser/Controller.java |    4 ++++
 src/com/android/browser/NavScreen.java  |    2 ++
 src/com/android/browser/PhoneUi.java    |   12 ++++++++++++
 src/com/android/browser/Tab.java        |    5 +++++
 4 files changed, 23 insertions(+), 0 deletions(-)

diff --git a/src/com/android/browser/Controller.java b/src/com/android/browser/Controller.java
index 2d716a8..4bdaf95 100644
--- a/src/com/android/browser/Controller.java
+++ b/src/com/android/browser/Controller.java
@@ -748,6 +748,10 @@ public class Controller
             mUploadHandler.onResult(Activity.RESULT_CANCELED, null);
             mUploadHandler = null;
         }
+        if (sThumbnailBitmap != null) {
+            sThumbnailBitmap.recycle();
+            sThumbnailBitmap = null;
+        }
         if (mTabControl == null) return;
         mUi.onDestroy();
         // Remove the current tab and sub window
diff --git a/src/com/android/browser/NavScreen.java b/src/com/android/browser/NavScreen.java
index 1d2114e..5df64de 100644
--- a/src/com/android/browser/NavScreen.java
+++ b/src/com/android/browser/NavScreen.java
@@ -162,6 +162,7 @@ public class NavScreen extends RelativeLayout
             } else {
                 mUiController.closeTab(tab);
             }
+            mTabViews.remove(tab);
         }
     }
 
@@ -238,6 +239,7 @@ public class NavScreen extends RelativeLayout
                 public void onClick(View v) {
                     if (tabview.isClose(v)) {
                         mScroller.animateOut(tabview);
+                        mTabViews.remove(tab);
                     } else if (tabview.isTitle(v)) {
                         switchToTab(tab);
                         mUi.getTitleBar().setSkipTitleBarAnimations(true);
diff --git a/src/com/android/browser/PhoneUi.java b/src/com/android/browser/PhoneUi.java
index 89eae70..86ffd94 100644
--- a/src/com/android/browser/PhoneUi.java
+++ b/src/com/android/browser/PhoneUi.java
@@ -74,6 +74,7 @@ public class PhoneUi extends BaseUi {
     @Override
     public void onDestroy() {
         hideTitleBar();
+        mAnimScreen.onDestroy();
     }
 
     @Override
@@ -514,6 +515,17 @@ public class PhoneUi extends BaseUi {
             mContent.setImageBitmap(mContentBitmap);
         }
 
+        public void onDestroy() {
+            if (mContentBitmap != null) {
+                mContentBitmap.recycle();
+                mContentBitmap = null;
+            }
+            if (mTitleBarBitmap != null) {
+                mTitleBarBitmap.recycle();
+                mTitleBarBitmap = null;
+            }
+        }
+
         private Bitmap safeCreateBitmap(int width, int height) {
             if (width <= 0 || height <= 0) {
                 Log.w(LOGTAG, "safeCreateBitmap failed! width: " + width
diff --git a/src/com/android/browser/Tab.java b/src/com/android/browser/Tab.java
index 026799b..96a379c 100644
--- a/src/com/android/browser/Tab.java
+++ b/src/com/android/browser/Tab.java
@@ -1224,6 +1224,11 @@ class Tab implements PictureListener {
      * Destroy the tab's main WebView and subWindow if any
      */
     void destroy() {
+        if (mCapture != null)
+        {
+            mCapture.recycle();
+            mCapture = null;
+        }
         if (mMainView != null) {
             dismissSubWindow();
             // save the WebView to call destroy() after detach it from the tab
-- 
1.7.4.1

